### ISSUE TITLE
Unavoidable psalm error MixedPropertyTypeCoercion

### ISSUE THREAD

**Issue by designermonkey** [vimeo/psalm#6354]
I'm not sure hw to get past this error: https://psalm.dev/r/b52287e5c5

```php
<?php declare(strict_types=1);

namespace Jorpo\Concerto\Application;

use Ds\Vector;

interface TransportObject
{
}

class TransportObjects
{
    /**
     * @var Vector<TransportObject>
     */
    private Vector $transportObjects;

    public function __construct(TransportObject ...$transportObjects)
    {
        $this->transportObjects = new Vector($transportObjects);
    }

    public function merge(TransportObjects $transportObjects): TransportObjects
    {
        $clone = clone $this;
        $clone->transportObjects = $clone->transportObjects->merge($transportObjects->transportObjects);

        return $clone;
    }
}
```

As far as I can see, the static analysis should be able to see that the specific implementation of `Vector` being passed here is type hinted to what it expects, therefore should not complain about possible `mixed` presence.

I don't know if it's a bug, or if anyone has advice on how to avoid this.

**Comment by weirdan** [comment]
Simplified: https://psalm.dev/r/19baf59fe3
Psalm fails to infer `T2` for `merge()`, failing to extract the type from `Vector<T>` when matching it to `iterable<int, T>`.
